### ISSUE TITLE
Support associated methods

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/cxx#51]
Currently only freestanding functions are supported by this crate, not methods.

I like the approach of using `self` in the foreign function parameter list to signify a method receiver:

```rust
#[cxx::bridge]
mod ffi {
    extern "C" {
        type C;
        fn f(self: &C);
    }

    extern "Rust" {
        type R;
        fn g(self: &R);
    }
}

struct R {...}

impl R {
    fn g(&self) {...}
}
```

This would expose the methods `r.g()` to C++ and `c.f()` to Rust.

The above syntax would require rustc 1.43 or newer for https://github.com/rust-lang/rust/pull/68728.

**Comment by myronahn** [comment]
Would it make sense to use the `&self` shorthand instead of `self: &C`?

**Comment by dtolnay** [comment]
That would be fine if there is only one extern type in the block.

**Comment by dtolnay** [comment]
Implemented in #121.

I filed #124 to follow up on `&self` and `&mut self` shorthand inside of extern blocks that contain exactly one extern type.
